### PR TITLE
Create solve competition api

### PR DESCRIPTION
**Pull request by vkgnosis**
Part of https://github.com/cowprotocol/services/issues/127 .

This is waiting on https://github.com/cowprotocol/services/pull/185 to know the exact interface.

Next PR will make use of the post endpoint in the driver.

### Test Plan

new unit tests

### CODE REVIEW

File: crates/orderbook/src/api/post_solver_competition.rs
Comment (nlordell) [inline crates/orderbook/src/api/post_solver_competition.rs:21]: Instead of adding auth code - we can just disallow `POST solver_competition` at the ingress level. We already connect to the orderbook using a cluster-local URL (so doesn't go through the ingress) and makes it so we don't need to manage secrets.

Comment (vkgnosis) [reply crates/orderbook/src/api/post_solver_competition.rs:21]: Not sure. Doing it only at the ingress level has the downside that an accidental misconfiguration could expose it. And we need to deal with another kubernetes configuration layer that I'm not familiar with. On the other hand it does make the code simpler and we don't want to keep this code for a long time anyway.

Comment (nlordell) [reply crates/orderbook/src/api/post_solver_competition.rs:21]: Personally, I would prefer it at an ingress level (I tend to prefer having generic access control configured at an ingress level as Nginx does a really good job with this).

That being said, code is already there, simple enough, and will hopefully go away in the near future, so I am fine with including it in the PR.

File: crates/model/src/solver_competition.rs
Comment (nlordell) [inline crates/model/src/solver_competition.rs:7]: Shouldn't this be a `Vec<(String, SolverSettlement)>` if we want to allow multiple settlements from the same solver (which, AFAIU, we currently do right - because of settlement merging?).

File: crates/model/src/solver_competition.rs
Comment (nlordell) [inline crates/model/src/solver_competition.rs:11]: If we use a `Vec` instead of a `HashMap` we should include `solver_name: String` here.

File: crates/orderbook/src/api/post_solver_competition.rs
Comment (nlordell) [inline crates/orderbook/src/api/post_solver_competition.rs:31]: Should this return a `No Content` or `Created` status code?

**Review by nlordell — COMMENTED** [review]
LGTM - only comment is that we should use a `Vec<SolverSettlement>` instead of a map so that:
- We can order by objective value 😄 (if you get the JSON its just a nicety that makes immediately seeing the winner and where you rank among the other settlements quicker).
- Allows for multiple settlements per solver. We technically have this with the merged settlements - although just uploading the "best" of the merged settlements also makes a lot of sense to me.

File: crates/orderbook/src/api/post_solver_competition.rs
Comment (vkgnosis) [reply crates/orderbook/src/api/post_solver_competition.rs]: Don't think it matters for our private temporary api so I didn't look into what the most proper  code is.

File: crates/model/src/solver_competition.rs
Comment (vkgnosis) [reply crates/model/src/solver_competition.rs:7]: Like I wrote above there, this is a placeholder until we merge the Openapi PR. Then I will make sure the types are right.

File: crates/orderbook/src/api/post_solver_competition.rs
Comment (vkgnosis) [reply crates/orderbook/src/api/post_solver_competition.rs]: Changed to Created

**Comment by vkgnosis** [comment]
Removed the authorization code. Planning on merging this and testing the ingress config on staging.

File: crates/orderbook/Cargo.toml
Comment (nlordell) [inline crates/orderbook/Cargo.toml:38]: I think this is no longer needed.

**Review by nlordell — APPROVED** [review]
Approving assuming that the type will be made to match the spec in a follow up PR (in order to verify the ingress configuration that will be needed).

**Pull request merged**
